### PR TITLE
Add a Twig helper for rendering figures

### PR DESCRIPTION
**Pull request by m-vo**
| Q                | A
| -----------------| ---
| Fixed issues     | #1508 
| Docs PR or issue | https://github.com/contao/docs/pull/468

*Updated description (20/07/08)*

---

### What's inside
This PR ships&hellip;
* &hellip; a Twig template + some helper functions (macros) to render a `Image\Studio\Figure` as a `<figure>` tag. This includes a picture + caption and is comparable to the `image.html5` template.
* &hellip; a Twig function to directly render figures from within templates. The function accepts an optional custom template and a set of options that used to configure the `FigureBuilder`.
* &hellip; as a bonus (idea &copy; by @ausi :slightly_smiling_face: ) a function to create `PictureConfiguration`s from within templates. This enables building figures with custom size configurations per template.

#### It enables the following scenarios:
1) Rendering the default template (from a Controller):
```php
$figure = $studio->createFigureBuilder->from($uuid)->setSize(5)->build();

$twig->render('@ContaoCore/Image/Studio/figure.html.twig', ['figure' => $figure]);
```

2) Rendering a custom template with a figure (using the 'figure' macro):
```php
$twig->render('my_template.html.twig', [
  'headline' => $headline,
  'my_figure' => $figure,
]);
```
```twig
{% import "@ContaoCore/Image/Studio/_macros.html.twig" as studio %}

<h1>{{ headline }}</h1>
<div class="container">
  {{ studio.figure(my_figure) }}
</div>
```

3) Rendering a figure from within a template :star2:  (&rarr; uses the default figure template):
```twig
{{ contao_figure(uuid, [200, 200, 'crop'], {
    enableLightBox: true,
    lightBoxGroupIdentifier: 'foo',
}) }}
```

4) Rendering a figure from within a template using a custom figure template:
```html
{{ contao_figure('path/to/an/image.png', '_mySize', {
    disableMetaData,
}, '@App\my_custom_template.html.twig') }}
```

5) Building a picture configuration on the fly (syntax equals definition from config):
```twig
{% set special_config = picture_config({
    width: 400,
    height: 400,
    resize_mode: 'crop',
    sizes: '0.75,1,1.5,2',
    items: [{
        width: 200,
        height: 100,
        media: '(max-width: 140px)',
    }]
}) %}

{{ contao_figure(id, special_config) }}
```

6) Building a figure template from building blocks:
```twig
{% import "@ContaoCore/Image/Studio/_macros.html.twig" as studio %}

<h3>Now look</h3>
<figure class="my_custom_stuff">
  <div class="caption-container">
    {{ studio.caption(figure) }}
  </div>
  {{ studio.picture(figure) }}
</div>
```

7) Overwriting options inside the template:
```twig
{% import "@ContaoCore/Image/Studio/_macros.html.twig" as studio %}

{{ studio.figure(uuid, {}, { 
    attr: { class: 'image-container' },
    caption_attr: { class: 'caption', data-foo: 'bar' }
} }}
```
This principle is borrowed from `symfony/form`. You can define options when creating a `Figure` but you can also specify them when building the template (overwriting the set ones). This makes the whole thing very flexible. Currently you can only define HTML attributes for the various tags but this can easily be extended by either the core or devs creating their own templates. Have a look in the macro file for a more detailed explanation how it is set up.

### Dependencies
Note: This PR is currently based on #1753. The relevant changes are in the following folders:
* `core-bundle\src\Twig\*`
* `core-bundle\views\Image\*`
* `core-bundle\Resources\config\services.yml`
* `core-bundle\tests\Twig\*`

(Commits starting from e76bf63295b6f5d2ff09fc17e74e463139bbb4ab)

### Todo
* [x] Finish templates
* [x] Tests
* [x] Rebase once #1753 is merged

### CODE REVIEW

**Comment by Toflar** [comment]
> Unfortunately similar code to retrieve the file/meta tags can be found multiple times in the code base now (e.g. `Controller::addImageToTemplate()` or the `picture` insert tag) - not sure if this is the most efficient way to do it.

Yeah, that's something that would be well suited for its own service anyway. I always cringe when I have to use the `Controller::addImageToTemplate()` without actually using a template...

**Comment by m-vo** [comment]
> Yeah, that's something that would be well suited for its own service anyway. 

It would be easy enough to put the `createPicture()` function into a service and call that instead (or rather return the data that needs to be set to the template). But we would need to discuss the API of such a service first (Array? DTO? Option to get it already rendered? ...).

> I  always cringe when I have to use the Controller::addImageToTemplate() without actually using a template...

Same :smile: I always feel like I don't know how to do it the right way.

**Comment by fritzmg** [comment]
Good stuff :). Could you may be add support for `{{ image(uuid, …) }}` too? ;)

**Comment by m-vo** [comment]
Notes to myself: 

* consider https://symfony.com/doc/current/templating/twig_extension.html#creating-lazy-loaded-twig-extensions
* escape possibly untrusted `$class` argument

**Comment by m-vo** [comment]
see #1537: if we decide we want to go this route, I'll postpone this PR until the other one is implemented

**Comment by m-vo** [comment]
depends on #1753

**Comment by m-vo** [comment]
This now uses the Image studio concept as a base. I updated the description accordingly.

**Comment by m-vo** [comment]
This one needs to be rebased once #1753 is merged. But otherwise it's imo ready to be reviewed.

We could also add some tests for the templates/macros itself (I'll have to find out what's the best way to do this first). But maybe that's something to do in a second step.

File: core-bundle/src/Twig/Runtime/PictureConfigurationRuntime.php
Comment (ausi) [inline core-bundle/src/Twig/Runtime/PictureConfigurationRuntime.php:65]: Should we throw an exception if `$configArray` has unused/invalid array keys? Would make it easier to spot typos, but would make it harder to write forwards-comaptible code (if new keys get added in the future).

**Comment by ausi** [comment]
It would be great to get autocompletion in the IDE for twig templates for the different configuration options. Do you know if something like that is possible?

File: core-bundle/src/Twig/Runtime/PictureConfigurationRuntime.php
Comment (m-vo) [reply core-bundle/src/Twig/Runtime/PictureConfigurationRuntime.php:65]: Yeah, I thought about this as well. It's probably a good idea to throw.

Perhaps we could use `PropertyAccess` as well here to set the values (after applying some minor transformations like `zoom` &rarr; `zoomLevel`). Then the code in the core would not need to know about possibly new config options in `PictureConfiguration`/`PictureConfigurationItem` and if you want to use the features you'd only have to add a dependency to a newer `contao/image` version. :thinking:

**Comment by m-vo** [comment]
> It would be great to get autocompletion in the IDE for twig templates for the different configuration options. Do you know if something like that is possible?

Unfortunately there seems to be no autocompletion capabilities for twig functions [other than the function name](https://symfony2-plugin.espend.de/languages/twig/index.html#extension) within the current PHPStorm Symfony plugin. You can get method completion for existing classes but that's cumbersome as well for the following reasons:
* you need to manually annotate your variables with `{# @var var \Type #}` (no inlining)
* return types [aren't recognized](https://github.com/Haehnchen/idea-php-symfony2-plugin/issues/1185), so no fluent interfaces either

Had a quick look into PHPStorm's advanced metadata, but there does not seem to be anything suitable Twig as well. :disappointed:

File: core-bundle/src/Twig/Runtime/PictureConfigurationRuntime.php
Comment (m-vo) [reply core-bundle/src/Twig/Runtime/PictureConfigurationRuntime.php:65]: I implemented said strategy in 89e114313402d8c5250bb69f981488655047cc51.

**Review by ausi — DISMISSED** [review]
Approved but needs to be rebased after #1753 got merged.

**Comment by m-vo** [comment]
Docs PR contao/docs#468 (draft)

**Comment by m-vo** [comment]
This is now rebased.

File: core-bundle/src/Twig/Extension/ImageExtension.php
Comment (m-vo) [inline core-bundle/src/Twig/Extension/ImageExtension.php:34]: In case you're wondering "Why the Runtime pattern?": 

1) It will only instantiate the runtime when the function is used in a rendered template (vs. everytime).
   see https://symfony.com/doc/current/templating/twig_extension.html#creating-lazy-loaded-twig-extensions
2) It's easier to register the functions under another name if you ever need to.

File: core-bundle/src/Twig/Runtime/FigureRendererRuntime.php
Comment (m-vo) [inline core-bundle/src/Twig/Runtime/FigureRendererRuntime.php:75]: I thought about adding another function `meta_data` (similar to `picture_configuration`) instead but figured that the creation of meta data is very cheap. There is no need to reuse it in the templates. 

So this way it can just be defined on the fly:
```twig
contao_figure(resource, {
  metaData: { alt: 'foo', caption: 'bar' },
  ...
})
```

**Review by ausi — DISMISSED** [review]

File: core-bundle/src/Resources/config/services.yml
Comment (leofeyer) [inline core-bundle/src/Resources/config/services.yml]: We should use the FQCN instead of a service ID if possible.

File: core-bundle/src/Twig/Extension/ImageExtension.php
Comment (leofeyer) [inline core-bundle/src/Twig/Extension/ImageExtension.php]: Maybe use `picture_config` instead? Would be shorter. (Your choice though.)

File: core-bundle/src/Resources/config/services.yml
Comment (m-vo) [reply core-bundle/src/Resources/config/services.yml]: You wouldn't want to autowire this, would you?

File: core-bundle/src/Twig/Extension/ImageExtension.php
Comment (m-vo) [reply core-bundle/src/Twig/Extension/ImageExtension.php]: I'm fine with both. I like the shorter name, but the longer one is consistent with the class name. :thinking: 

/cc @ausi

Comment (ausi) [reply core-bundle/src/Twig/Extension/ImageExtension.php]: +1 for `picture_config`

Comment (m-vo) [reply core-bundle/src/Twig/Extension/ImageExtension.php]: Shortened in c287ab880953bb836a0cc653f64da57ae9b41a60

**Comment by m-vo** [comment]
I made the size parameter an integral part of the `contao_figure` signature in 7b5ea547e70c78516aa30bd17e481c56f7cb3550.

The size parameter is needed 99% of the time. And this allows an even more compact way of writing it:
```twig
{{ contao_figure('/path/to/my/image.png', '_my_size' }}
```

see https://github.com/contao/contao/pull/1937#issuecomment-659440194

(failing tests are unrelated)

**Review by ausi — DISMISSED** [review]

File: core-bundle/src/Resources/config/services.yml
Comment (leofeyer) [reply core-bundle/src/Resources/config/services.yml]: I don't know what you mean. 😄 We want to use FQCNs for all services in the future as recommended by Symfony.

File: core-bundle/tests/Twig/Runtime/FigureRendererRuntimeTest.php
Comment (leofeyer) [inline core-bundle/tests/Twig/Runtime/FigureRendererRuntimeTest.php]: This will also expect exceptions in the `getRuntime()` method. Intended?

**Review by leofeyer — CHANGES_REQUESTED** [review]
You may want to re-read the comments in the `_macros.html.twig` file. They seem a little inconsistent, because some variables and class names are wrapped in backticks and some are not. IMHO backticks do not have any effect in Twig comments at all, have they?

File: core-bundle/tests/Twig/Runtime/FigureRendererRuntimeTest.php
Comment (m-vo) [reply core-bundle/tests/Twig/Runtime/FigureRendererRuntimeTest.php]: Ah, yes you're right. Got me again.

**Comment by m-vo** [comment]
> IMHO backticks do not have any effect in Twig comments at all, have they?

I removed them in 2f2bfcf9988e7201127c672a60de732e6bbda53d and only left single quotes at one place to for better readability.

File: core-bundle/tests/Twig/Runtime/FigureRendererRuntimeTest.php
Comment (m-vo) [reply core-bundle/tests/Twig/Runtime/FigureRendererRuntimeTest.php]: Fixed in a70beef2aba5099db37f8da41e03634608baf149

File: core-bundle/src/Resources/config/services.yml
Comment (m-vo) [reply core-bundle/src/Resources/config/services.yml]: Changed to FQCN in ff2538a40e248d02da8df72f424097c168beaf86

File: core-bundle/src/Image/Studio/FigureBuilder.php
Comment (leofeyer) [inline core-bundle/src/Image/Studio/FigureBuilder.php:274]: Don't you have to adjust the property phpDoc as well? Currently it is `@var mixed|null`, which is definitely wrong, because `mixed` already includes `null`. So either you remove the `@param` annotation completely (then `mixed` is implied) or you adjust the `@var` annotation accordingly.

Comment (m-vo) [reply core-bundle/src/Image/Studio/FigureBuilder.php:274]: Ah I missed that. Fixed in 74310e3eabc7a8243fe6c47246d31a37e50e46f9.

**Review by leofeyer — APPROVED** [review]

**Review by ausi — APPROVED** [review]

**Comment by leofeyer** [comment]
Thank you @m-vo.

**Pull request merged**
